extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/logical/LogicalStageBuilder.java

```diff
@@ -0,0 +1,273 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class LogicalStageBuilder
+{
+  private PlannerContext plannerContext;
+
+  public LogicalStageBuilder(PlannerContext plannerContext2)
+  {
+    this.plannerContext = plannerContext2;
+  }
+
+  class StageMaker
+  {
+    public final AtomicInteger stageIdSeq = new AtomicInteger(0);
+
+    StageDefinition makeScanStage(
+        VirtualColumns virtualColumns,
+        RowSignature signature,
+        List<InputSpec> inputs,
+        DimFilter dimFilter)
+    {
+      ScanQuery s = Druids.newScanQueryBuilder()
+          .dataSource(IRRELEVANT)
+          .intervals(QuerySegmentSpec.ETERNITY)
+          .filters(dimFilter)
+          .virtualColumns(virtualColumns)
+          .columns(signature.getColumnNames())
+          .columnTypes(signature.getColumnTypes())
+          .build();
+      ScanQueryFrameProcessorFactory scanProcessorFactory = new ScanQueryFrameProcessorFactory(s);
+      StageDefinitionBuilder sdb = StageDefinition.builder(stageIdSeq.incrementAndGet())
+          .inputs(inputs)
+          .signature(signature)
+          .shuffleSpec(MixShuffleSpec.instance())
+          .processorFactory(scanProcessorFactory);
+      return sdb.build(getIdForBuilder());
+    }
+
+    private String getIdForBuilder()
+    {
+      String dartQueryId = plannerContext.queryContext().getString(QueryContexts.CTX_DART_QUERY_ID);
+      if (dartQueryId != null) {
+        return dartQueryId;
+      }
+      return plannerContext.getSqlQueryId();
+    }
+  }
+
+  public class AbstractLogicalStage implements LogicalStage
+  {
+    protected final List<InputSpec> inputSpecs;
+    protected final RowSignature signature;
+    protected final List<LogicalStage> inputStages;
+
+    public AbstractLogicalStage(RowSignature signature, List<InputSpec> inputs, List<LogicalStage> inputVertices)
+    {
+      this.inputSpecs = inputs;
+      this.signature = signature;
+      this.inputStages = inputVertices;
+    }
+
+    @Override
+    public StageDefinition buildCurrentStage(StageMaker stageMaker)
+    {
+      throw DruidException.defensive("This should have been implemented - or not reach this point!");
+    }
+
+    @Override
+    public LogicalStage extendWith(DruidNodeStack stack)
+    {
+      return null;
+    }
+
+    @Override
+    public final QueryDefinition build()
+    {
+      return QueryDefinition.create(buildStageDefinitions(new StageMaker()), plannerContext.queryContext());
+    }
+
+    @Override
+    public final List<StageDefinition> buildStageDefinitions(StageMaker stageMaker)
+    {
+      List<StageDefinition> ret = new ArrayList<>();
+      if (!inputStages.isEmpty()) {
+        throw DruidException.defensive("Not yet supported");
+      }
+      ret.add(buildCurrentStage(stageMaker));
+      return ret;
+    }
+  }
+
+  public class RootStage extends AbstractLogicalStage
```

Comment:
Could use some definition of what "root stage" means.

Reply:
now I feel like the usage of the `Root` keyword is misleading here...I've renamed it to `ReadStage` ; other alternatives were:
* `ReadStage`
* `InputStage` - second best
* `ScanStage` -  the presence of `ScanQuery` and other usages of `scan` makes it a possibly confusing one
* `PrepareStage` - sounds too generic

Reply:
`LeafStage`? `ReadStage` is fine too.